lib/spack/spack/package.py

```diff
@@ -1444,13 +1444,14 @@ def do_install(self, **kwargs):
                 dep.package.do_install(**dep_kwargs)
 
         # Then, install the compiler if it is not already installed.
-        if install_deps:
-            tty.debug('Boostrapping {0} compiler for {1}'.format(
-                self.spec.compiler, self.name
-            ))
-            comp_kwargs = kwargs.copy()
-            comp_kwargs['explicit'] = False
-            self.bootstrap_compiler(**comp_kwargs)
+        if install_deps or not explicit:
```

Comment:
@becker33 

I'm having trouble understanding the conditions when this holds. It feels like this is addressing the problem from the perspective of how the function is called rather than from the perspective of just deciding whether the compiler should be installed. I think it would make more sense to get rid of this condition and have the entire block be:

```
if spack.config.get('config:install_missing_compilers', False):
  tty.debug('Boostrapping {0} compiler for {1}'.format(
    self.spec.compiler, self.name
  ))
  comp_kwargs = kwargs.copy()
  comp_kwargs['explicit'] = False
  self.bootstrap_compiler(**comp_kwargs)
```

I think we might as well do that if we are going to conditionally ignore the "mandate" of `install_deps` (i.e. if we are willing to install the compiler even when `install_deps == False` in some cases, we might as well ignore it always).

Reply:
If install_deps is False and explicit is False, that means that we're installing a dependency of something that was installed with install_deps True. We turn install_deps off for all of the dependencies so that we can use an iterative algorithm instead of a recursive one. But the one type of dependency we don't account for in that iterative algorithm is the compiler.

The other option is to install the compiler before the dependencies. However, that will get confusing in cases in which the dependencies have their own compilers. Say I'm installing `openssl %intel ^ zlib%gcc`. I want the install order to be `gcc`, `zlib`, `intel`, `openssl`. That's why I don't think reordering is the right option.

Reply:
I agree with what you say and think I am familiar with the background logic like

> We turn install_deps off for all of the dependencies so that we can use an iterative algorithm instead of a recursive one

But I don't see how my suggestion conflicts with any of those statements. At this point we have to decide whether the compiler is available and if it is not we have to install it (since right after these lines, we start trying to install the package itself). I think we can skip considerations of how this function is called, and just decide based on the config and what the package needs, whether to proceed with trying to install the compiler.

Minor side note, I just noticed the error message contains a typo: `Boostrapping`

Reply:
What about the case in which we're installing an explicit package with `install_deps` false. That means we specifically requested on the command line not to install dependencies. I think in that case we shouldn't install the compiler, even if `install_missing_compilres` is configured true.

Reply:
That's true but given that the current `if` statement already softened that rule:

> `if install_deps or not explicit:`

I figured why not do away with that entirely. As we discuss it though perhaps both approaches are confusing and suggest a change to the API.

* One possibility is to add another argument in addition to `install_deps` called `install_compiler_deps`, which would feel redundant but I think would be easier to explain.
* I think another option is to redefine the meaning of `install_deps` to say that we skip all dependencies except for compiler dependencies of the package being considered (this is similar to your approach and also is somewhat hypocritical since it reflects understanding of how the method is currently invoked, but I think would more-directly codify that convention).

Do either of those appeal? I'm certainly open to other suggestions as well. I'm a bit concerned that the API as it is right now is a bit confusing.

Reply:
Ok I had another idea. What about switching to not use recursion for the compilers. For each node in the DAG, check whether to install compilers and run bootstrap compilers if necessary.

I'm pretty sure that's the right solution. It preserves the same logic while being much more understandable. I'm running some tests now to confirm it works as I would expect and I'll push new code in the next hour.